src/client/ws/manager.rs

```diff
@@ -0,0 +1,278 @@
+	requests: FnvHashMap<RequestId, Kind>,
+	/// Reverse lookup, to find a request ID in constant time by `subscription ID` instead of looking through all requests.
+	subscriptions: HashMap<SubscriptionId, RequestId>,
+}
+
+impl RequestManager {
+	pub fn new() -> Self {
+		Self { requests: FnvHashMap::default(), subscriptions: HashMap::new() }
+	}
+
+	/// Tries to insert a new pending request.
+	pub fn insert_pending_method_call(
+		&mut self,
+		id: u64,
+		send_back: PendingRequestOneshot,
+	) -> Result<(), PendingRequestOneshot> {
+		if let Entry::Vacant(v) = self.requests.entry(id) {
+			v.insert(Kind::PendingMethodCall(send_back));
+			Ok(())
+		} else {
+			Err(send_back)
+		}
+	}
+
+	/// Tries to inserts a new pending subscription.
+	pub fn insert_pending_subscription(
+		&mut self,
+		id: RequestId,
+		send_back: PendingSubscriptionOneshot,
+		unsubscribe_method: UnsubscribeMethod,
+	) -> Result<(), PendingSubscriptionOneshot> {
+		if let Entry::Vacant(v) = self.requests.entry(id) {
+			v.insert(Kind::PendingSubscription((send_back, unsubscribe_method)));
+			Ok(())
+		} else {
+			Err(send_back)
+		}
+	}
+
+	/// Inserts a new active subscription
+	pub fn insert_active_subscription(
+		&mut self,
+		request_id: RequestId,
+		subscription_id: SubscriptionId,
+		send_back: SubscriptionSink,
+		unsubscribe_method: String,
+	) -> Result<(), SubscriptionSink> {
+		if let (Entry::Vacant(request), Entry::Vacant(subscription)) =
+			(self.requests.entry(request_id), self.subscriptions.entry(subscription_id))
+		{
+			request.insert(Kind::Subscription((send_back, unsubscribe_method)));
+			subscription.insert(request_id);
+			Ok(())
+		} else {
+			Err(send_back)
+		}
+	}
+
+	/// Complete a pending subscription.
+	pub fn complete_pending_subscription(
+		&mut self,
+		request_id: RequestId,
+	) -> Option<(PendingSubscriptionOneshot, UnsubscribeMethod)> {
+		match self.requests.entry(request_id) {
+			Entry::Occupied(request) if matches!(request.get(), Kind::PendingSubscription(_)) => {
+				let (_req_id, kind) = request.remove_entry();
+				if let Kind::PendingSubscription(send_back) = kind {
+					Some(send_back)
+				} else {
+					unreachable!();
+				}
+			}
+			_ => None,
+		}
+	}
+
+	/// Complete a pending request.
+	pub fn complete_pending_method_call(&mut self, request_id: RequestId) -> Option<PendingRequestOneshot> {
+		match self.requests.entry(request_id) {
+			Entry::Occupied(request) if matches!(request.get(), Kind::PendingMethodCall(_)) => {
+				let (_req_id, kind) = request.remove_entry();
+				if let Kind::PendingMethodCall(send_back) = kind {
+					Some(send_back)
+				} else {
+					unreachable!();
+				}
+			}
+			_ => None,
+		}
+	}
+
+	/// Remove an active subscription.
+	pub fn remove_active_subscription(
+		&mut self,
+		request_id: RequestId,
+		subscription_id: SubscriptionId,
+	) -> Option<(SubscriptionSink, UnsubscribeMethod)> {
+		match (self.requests.entry(request_id), self.subscriptions.entry(subscription_id)) {
```

Comment:
Entry API takes ownership of the key, could do:

```rust
let request = matches!(self.request.get(&request_id), Some(Kind::Subscription(_));
let sub = matches!(self.request.get(&request_id), Some(_);
if request && sub {
   // remove request and sub.
}
```

Still have to do two lookups though :(

Reply:
//cc @dvdplm thoughts?